arachne-storage/src/main/java/com/odysseusinc/arachne/storage/service/JcrContentStorageServiceImpl.java

```diff
@@ -0,0 +1,261 @@
+import javax.jcr.NodeIterator;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.query.Query;
+import javax.jcr.query.QueryManager;
+import javax.jcr.query.QueryResult;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.Table;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.jackrabbit.JcrConstants;
+import org.apache.jackrabbit.commons.JcrUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.convert.ConversionService;
+import org.springframework.stereotype.Service;
+import org.springmodules.jcr.JcrTemplate;
+
+@Service
+public class JcrContentStorageServiceImpl implements ContentStorageService {
+
+    public static String PATH_SEPARATOR = "/";
+    public static String ENTITY_FILES_DIR = "entities";
+
+    public static String JCR_CONTENT_TYPE = "jcr:contentType";
+    public static String JCR_AUTHOR = "jcr:author";
+
+    protected JcrTemplate jcrTemplate;
+    protected ConversionService conversionService;
+    protected EntityManagerFactory entityManagerFactory;
+
+    @Autowired
+    public JcrContentStorageServiceImpl(JcrTemplate jcrTemplate,
+                                        ConversionService conversionService,
+                                        EntityManagerFactory entityManagerFactory) {
+
+        this.jcrTemplate = jcrTemplate;
+        this.conversionService = conversionService;
+        this.entityManagerFactory = entityManagerFactory;
+    }
+
+    public String getLocationForEntity(String entityTableName, Serializable entityIdentifier, List<String> additionalPathParts) {
+
+        List<String> pathParts = new ArrayList<>(Arrays.asList(ENTITY_FILES_DIR, entityTableName, entityIdentifier.toString()));
+
+        pathParts.addAll(additionalPathParts);
+
+        return PATH_SEPARATOR + pathParts.stream()
+                .filter(part -> StringUtils.isNotBlank(part) && !part.equals(PATH_SEPARATOR))
+                .collect(Collectors.joining(PATH_SEPARATOR));
+    }
+
+    public String getLocationForEntity(Object domainObject, List<String> additionalPathParts) {
+
+        Table entityTable = domainObject.getClass().getAnnotation(Table.class);
+        String entityId = String.valueOf(entityManagerFactory.getPersistenceUnitUtil().getIdentifier(domainObject));
+
+        return getLocationForEntity(entityTable.name(), entityId, additionalPathParts);
+    }
+
+    @Override
+    public ArachneFileSourced getFileByPath(String absoulteFilename) {
+
+        return (ArachneFileSourced) jcrTemplate.execute(session -> {
+
+            Node fileNode = session.getNode(absoulteFilename);
+            return getFile(fileNode);
+        });
+    }
+
+    public ArachneFileSourced getFileByIdentifier(String identifier) {
+
+        return (ArachneFileSourced) jcrTemplate.execute(session -> {
```

Comment:
Dont you think it would be better to wrap all calls to foreign library with wrapper like this:
`<T> T execute(callback, Class<T>)`